File: gateway/src/main/java/bank/internalgateway/gateway/identity/IdentityEnvelopeService.java
package bank.internalgateway.gateway.identity;

import bank.internalgateway.dsl.BenchmarkRouteRegistry;
import bank.internalgateway.dsl.CompiledBenchmarkModule;
import bank.internalgateway.gateway.config.GatewayProperties;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.stereotype.Service;

import javax.crypto.SecretKey;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Date;
import java.util.Map;
import java.util.UUID;

@Service
public class IdentityEnvelopeService {

    private final SecretKey secretKey;
    private final String fallbackIssuer;
    private final int fallbackTtlSeconds;
    private final String fallbackEvidenceId;
    private final BenchmarkRouteRegistry benchmarkRouteRegistry;

    public IdentityEnvelopeService(GatewayProperties properties, BenchmarkRouteRegistry benchmarkRouteRegistry) {
        this.secretKey = Keys.hmacShaKeyFor(properties.envelopeSecret().getBytes(StandardCharsets.UTF_8));
        GatewayProperties.Envelope envelope = properties.envelope();
        this.fallbackIssuer = envelope != null && envelope.issuer() != null ? envelope.issuer() : "internal-gateway";
        this.fallbackTtlSeconds = envelope != null && envelope.ttlSeconds() != null && envelope.ttlSeconds() > 0
                ? envelope.ttlSeconds()
                : 30;
        this.fallbackEvidenceId = envelope != null && envelope.businessControlEvidenceId() != null
                ? envelope.businessControlEvidenceId()
                : "poc-stub-passed";
        this.benchmarkRouteRegistry = benchmarkRouteRegistry;
    }

    public String createBankUserEnvelope(String subjectId, String organizationId, String correlationId) {
        return createBankUserEnvelope(subjectId, organizationId, correlationId, null);
    }

    public String createBankUserEnvelope(
            String subjectId,
            String organizationId,
            String correlationId,
            String evidenceIdOverride) {
        CompiledBenchmarkModule.EnvelopePolicy policy = envelopePolicy();
        String evidenceId = evidenceIdOverride != null && !evidenceIdOverride.isBlank()
                ? evidenceIdOverride
                : policy.businessControlEvidenceId();
        Instant now = Instant.now();
        return Jwts.builder()
                .issuer(policy.issuer())
                .subject(subjectId)
                .claim("organizationId", organizationId)
                .claim("correlationId", correlationId)
                .claim("operationId", UUID.randomUUID().toString())
                .claim("businessControlEvidenceId", evidenceId)
                .issuedAt(Date.from(now))
                .expiration(Date.from(now.plusSeconds(policy.ttlSeconds())))
                .signWith(secretKey)
                .compact();
    }

    public String createDeliveryEnvelope(Map<String, Object> claims) {
        CompiledBenchmarkModule.EnvelopePolicy policy = envelopePolicy();
        Instant now = Instant.now();
        var builder = Jwts.builder()
                .issuer(policy.issuer())
                .claim("deliveryMode", "signed-delivery-envelope")
                .issuedAt(Date.from(now))
                .expiration(Date.from(now.plusSeconds(policy.ttlSeconds())))
                .signWith(secretKey);
        claims.forEach(builder::claim);
        return builder.compact();
    }

    private CompiledBenchmarkModule.EnvelopePolicy envelopePolicy() {
        try {
            return benchmarkRouteRegistry.currentModule().envelopePolicy();
        } catch (IllegalStateException ex) {
            return new CompiledBenchmarkModule.EnvelopePolicy(
                    fallbackIssuer, fallbackTtlSeconds, java.util.List.of(), fallbackEvidenceId);
        }
    }
}

File: dotnet-gateway/src/InternalGateway.DotNet/Identity/IdentityEnvelopeService.cs
using System.Security.Cryptography;
using System.Text;
using Microsoft.Extensions.Options;
using Microsoft.IdentityModel.JsonWebTokens;
using Microsoft.IdentityModel.Tokens;

namespace InternalGateway.DotNet.Identity;

public sealed class IdentityEnvelopeService
{
    private readonly GatewayOptions _options;
    private readonly OffersRouteRegistryAccessor _offers;
    private readonly SigningCredentials _credentials;
    private readonly JsonWebTokenHandler _handler = new();

    public IdentityEnvelopeService(IOptions<GatewayOptions> options, OffersRouteRegistryAccessor offers)
    {
        _options = options.Value;
        _offers = offers;
        var keyBytes = Encoding.UTF8.GetBytes(_options.EnvelopeSecret);
        if (keyBytes.Length < 32)
        {
            Array.Resize(ref keyBytes, 32);
        }

        _credentials = new SigningCredentials(
            new SymmetricSecurityKey(keyBytes),
            SecurityAlgorithms.HmacSha256);
    }

    public string CreateBankUserEnvelope(
        string subjectId,
        string organizationId,
        string correlationId,
        string? evidenceIdOverride = null)
    {
        var policy = TryEnvelopePolicy();
        var evidenceId = !string.IsNullOrWhiteSpace(evidenceIdOverride)
            ? evidenceIdOverride
            : policy.BusinessControlEvidenceId;
        var now = DateTime.UtcNow;
        var descriptor = new SecurityTokenDescriptor
        {
            Issuer = policy.Issuer,
            Subject = new System.Security.Claims.ClaimsIdentity(
            [
                new System.Security.Claims.Claim(JwtRegisteredClaimNames.Sub, subjectId),
                new System.Security.Claims.Claim("organizationId", organizationId),
                new System.Security.Claims.Claim("correlationId", correlationId),
                new System.Security.Claims.Claim("operationId", Guid.NewGuid().ToString("D")),
                new System.Security.Claims.Claim("businessControlEvidenceId", evidenceId)
            ]),
            NotBefore = now,
            Expires = now.AddSeconds(policy.TtlSeconds),
            SigningCredentials = _credentials
        };
        return _handler.CreateToken(descriptor);
    }

    public string CreateDeliveryEnvelope(IReadOnlyDictionary<string, object?> claims)
    {
        var policy = TryEnvelopePolicy();
        var now = DateTime.UtcNow;
        var identity = new System.Security.Claims.ClaimsIdentity();
        identity.AddClaim(new System.Security.Claims.Claim("deliveryMode", "signed-delivery-envelope"));
        foreach (var (key, value) in claims)
        {
            if (value is not null)
            {
                identity.AddClaim(new System.Security.Claims.Claim(key, Convert.ToString(value) ?? string.Empty));
            }
        }

        var descriptor = new SecurityTokenDescriptor
        {
            Issuer = policy.Issuer,
            Subject = identity,
            NotBefore = now,
            Expires = now.AddSeconds(policy.TtlSeconds),
            SigningCredentials = _credentials
        };
        return _handler.CreateToken(descriptor);
    }

    private Dsl.EnvelopePolicy TryEnvelopePolicy()
    {
        try
        {
            return _offers.Registry.CurrentModule.EnvelopePolicy;
        }
        catch
        {
            return new Dsl.EnvelopePolicy(
                _options.Envelope.Issuer,
                _options.Envelope.TtlSeconds,
                [],
                _options.Envelope.BusinessControlEvidenceId);
        }
    }
}

/// <summary>Late-bound accessor so identity can be constructed before registry load completes in DI.</summary>
public sealed class OffersRouteRegistryAccessor
{
    public Dsl.OffersRouteRegistry Registry { get; set; } = null!;
}
